libvast/include/vast/cast.hpp

```diff
@@ -243,6 +257,647 @@ struct cast_helper<record_type, record_type> {
+    die("unimplemented");
+  }
+};
+
+template <>
+struct cast_helper<double_type, int64_type> {
+  static auto can_cast(const double_type&, const int64_type&) noexcept
+    -> caf::expected<void> {
+    return {};
+  }
+
+  static auto cast_value(const double_type&, double value,
+                         const int64_type&) noexcept -> caf::expected<int64_t> {
+    if (std::isinf(value))
+      return caf::make_error(ec::convert_error,
+                             fmt::format("unable to convert {} into int64. The "
+                                         "input must be a finite number",
+                                         value));
+    if (double i; std::modf(value, &i))
+      return caf::make_error(ec::convert_error,
+                             fmt::format("unable to convert {} into int64. "
+                                         "Fractional part not allowed",
+                                         value));
+
+    if (value > static_cast<double>(std::numeric_limits<int64_t>::max())
+        or value < static_cast<double>(std::numeric_limits<int64_t>::min())) {
+      return caf::make_error(ec::convert_error,
+                             fmt::format("unable to convert {} into int64. The "
+                                         "value is out of int64_t range",
+                                         value));
+    }
+    return static_cast<int64_t>(value);
+  }
+
+  static auto
+  cast(const double_type&, std::shared_ptr<type_to_arrow_array_t<double_type>>,
+       const int64_type&) noexcept
+    -> std::shared_ptr<type_to_arrow_array_t<int64_type>> {
+    die("unimplemented");
+  }
+};
+
+template <>
+struct cast_helper<uint64_type, double_type> {
+  static auto can_cast(const uint64_type&, const double_type&) noexcept
+    -> caf::expected<void> {
+    return {};
+  }
+
+  static auto cast_value(const uint64_type&, uint64_t value,
+                         const double_type&) noexcept -> caf::expected<double> {
+    constexpr auto max_loseless_integer
+      = (uint64_t{1} << std::numeric_limits<double>::digits) - 1;
+    if (value > max_loseless_integer) {
+      return caf::make_error(ec::convert_error,
+                             "unable to convert {} into a double. The "
+                             "conversion may result in loss of precision");
+    }
+    return static_cast<double>(value);
+  }
+
+  static auto
+  cast(const uint64_type&, std::shared_ptr<type_to_arrow_array_t<uint64_type>>,
+       const double_type&) noexcept
+    -> std::shared_ptr<type_to_arrow_array_t<double_type>> {
+    die("unimplemented");
+  }
+};
+
+template <>
+struct cast_helper<double_type, uint64_type> {
+  static auto can_cast(const double_type&, const uint64_type&) noexcept
+    -> caf::expected<void> {
+    return {};
+  }
+
+  static auto
+  cast_value(const double_type&, double value, const uint64_type&) noexcept
+    -> caf::expected<uint64_t> {
+    if (std::isinf(value))
```

Comment:
```suggestion
    if (not std::isfinite(value))
```

That way you catch inf, -inf, and NaN

Reply:
Good catch!